File: python/questions/isSymmetric.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        def cLR(l: Optional[TreeNode], r: Optional[TreeNode]) -> bool:
            if l and r:
                return cLR(l.left, r.right) and cLR(l.right, r.left) if l.val==r.val else False
            if l is None and r is None:
                return True
            return False
        return cLR(root.left, root.right) if root else True

File: cpp/questions/isSymmetric.cpp
#include <functional>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        std::function<bool(TreeNode*, TreeNode*)> cLR = [&](TreeNode* l, TreeNode* r) -> bool {
            if (l && r) {
                return l->val==r->val ? cLR(l->left, r->right) && cLR(l->right, r->left) : false;
            }
            if (l==nullptr && r == nullptr) {
                return true;
            }
            return false;
        };
        return root ? cLR(root->left, root->right) : true;
    }
};
